src/publisher.js

```diff
@@ -130,21 +155,59 @@ Publisher.prototype.publish = function () {
 
 					http(config, function (error, response) {
 						if (!error && (response.statusCode >= 200 && response.statusCode < 300)) {
-							deferred.resolve();
+							publishDeferred.resolve(data);
 						} else {
 							if (error != null) {
-								deferred.reject(error);
+								publishDeferred.reject(error);
 							} else {
-								deferred.reject(new Error('Unable to publish Pact to Broker: ' + response.statusCode));
+								publishDeferred.reject(new Error('Unable to publish Pact to Broker: ' + response.statusCode));
 							}
 						}
 					});
 
-					return deferred.promise;
+					return publishDeferred.promise;
 				}, function(err) {
 					return q.reject(err);
 				})
+				.then(function(data) {
+					if (!options.tags.length) {
+						deferred.resolve();
+						return deferred.promise;
+					}
+
+					return q.all(_.map(options.tags, function(tag) {
+						var tagDeferred = q.defer();
+						var tagUrl = constructTagUrl(options, tag, data)
 
+						var config = {
+							uri: tagUrl,
+							method: 'PUT',
+							headers: { 'Content-Type': 'application/json' },
+							json: true,
+							auth: auth
+						};
+
+						http(config, function (error, response) {
+							if (!error && response.statusCode >= 200 && response.statusCode < 300) {
+								tagDeferred.resolve();
+							} else {
+								if (error != null) {
+									tagDeferred.reject(error);
+								} else {
+									tagDeferred.reject(new Error('Could not tag Pact with tag "' + tag + '": ' + response.statusCode));
+								}
+							}
+						});
+
+						return tagDeferred.promise;
+					}))
+				}, function(err) {
+					return q.reject(err);
+				})
+				.then(function() {
+					deferred.resolve();
+					return deferred.promise;
```

Comment:
There must be a way to flatten all of these promises. I might be a Bluebird fanboy (for no good reason other than I've used it(, but I'd find it hard to believe q can't do it too. No doubt it would have influenced Bluebirds design. I feel this would make all of this easier to read (to be fair, previous to this PR it wasn't pretty either and I'm probably at least in part to blame).